ui/components/ui/box/box.js

```diff
@@ -104,54 +88,113 @@ export const ValidBorderColors = [
-    [`${baseClass}--${type}-right-${right}`]: isValidValue(type, right),
-    [`${baseClass}--${type}-bottom-${bottom}`]: isValidValue(type, bottom),
-    [`${baseClass}--${type}-left-${left}`]: isValidValue(type, left),
-    // As long as an array of length >= 2 has been provided, the first number
-    // will always be for the top value.
-    [`${baseClass}--${type}-top-${arr?.[0]}`]: hasAtLeastTwo,
-    // As long as an array of length >= 2 has been provided, the second number
-    // will always be for the right value.
-    [`${baseClass}--${type}-right-${arr?.[1]}`]: hasAtLeastTwo,
-    // If an array has 2 values, the first number is the bottom value. If
-    // instead if has 3 or more values, the third number will be the bottom.
-    [`${baseClass}--${type}-bottom-${arr?.[2]}`]: hasAtLeastThree,
-    [`${baseClass}--${type}-bottom-${arr?.[0]}`]: isHorizontalAndVertical,
-    // If an array has 2 or 3 values, the second number will be the left value
-    [`${baseClass}--${type}-left-${arr?.[1]}`]:
-      isHorizontalAndVertical || isTopHorizontalAndBottom,
-    // If an array has 4 values, the fourth number is the left value
-    [`${baseClass}--${type}-left-${arr?.[3]}`]: isAllFour,
-  };
+function generateClassNames(baseClass, type, value, validatorFn) {
+  // if value does not exist return null
+  if (!value) {
+    return null;
+  }
+  // initiate classObject to be returned to classnames package
+  let classesObject = {};
+  // check if prop is not an array prop
+  let singleDigit = Array.isArray(value) ? undefined : value;
+  // check if single digit exists or if array has only one item
+  if (singleDigit || value.length === 1) {
+    // if it is an array with only one item assign it to singleDigit
+    if (value.length === 1) {
+      singleDigit = value[0];
+    }
+    // add base style without any className breakpoint prefixes to classObject
+    classesObject = {
+      ...classesObject,
+      [`${baseClass}--${type}-${singleDigit}`]: validatorFn
+        ? validatorFn(type, singleDigit)
+        : true,
+    };
+  } else {
+    // for array prop with more than one item
+    for (let i = 0; i < value.length; i++) {
+      // omit any null values to skip breakpoints
+      if (value[i] !== null) {
+        // first value is always the base value so don't apply any breakpoint prefixes
+        if (i === 0) {
+          classesObject = {
+            ...classesObject,
+            [`${baseClass}--${type}-${value[i]}`]: validatorFn
+              ? validatorFn(type, value[i])
+              : true,
+          };
+        } else {
+          // apply breakpoint prefixes according to index in array [base(no prefix), sm:, md:, lg:]
+          classesObject = {
+            ...classesObject,
+            [`${baseClass}--${BREAKPOINTS[i]}:${type}-${value[i]}`]: validatorFn
+              ? validatorFn(type, value[i])
+              : true,
+          };
+        }
+      }
```

Comment:
This looping concerns me, given the potential number of calls we'll have to this method. 

Could we perhaps make some optimizations based on what we know. AKA we have only 4 breakpoints, base and 3 others. We also know that at this point in the code we have at least length 2 otherwise we wouldn't be in the else block. 

So could we not optimize this by way of a switch statement? 

```js
// add the base class.
switch(value.length) {
  case 4:
    // add all classes
 case 3:
   // add just sm/md
 case 2:
  // add just sm
 default:
  // do nothing? 
}
```
doing something like this reduces time complexity of the function itself to linear.